RFC for Microsoft.PowerShell.Archive v2.0.0 (#117)
PowerShell/Microsoft.PowerShell.Archive: rfc/RFCNNNN-Archive-Module-v2.md

This RFC proposes new features and changes for the existing Microsoft.PowerShell.Archive module.
The goal for the Archive module is to build a high-performing and maintainable module that offers high utility and works cross-platform (especially with regard to file paths).

Currently, the archive module has a number of limitations.
First, the module supports the zip32 format only.
There is an opportunity to support additional supports by taking advanatage of new .NET APIs.
Second, the module has limited performance compared to other archive software.
Writing the next version of the module in C# instead of PowerShell Script is expected to improve the overall performance of the module.
The module has limited cross-platform support because archive entries are written in an OS-specifc way due to different characters being used as path seperators in different OSs.
This makes it difficult for Unix-based OS users to use archives compressed on a Windows computer or vice versa.

There are also a number of usability issues reported by users.
For example, there are issues with wildcard characters in paths.
Error reporting in some circumstances can be improved and more descriptive.

> Compatability with other archive software can also be improved as there are cases where an archive program may not recognize an archive produced by this module as valid.

Reviewer: ```suggestion
Compatibility with other archive software can also be improved as there are cases where an archive program may not recognize an archive produced by this module as valid.

```

---
RFC:
Author: Abdullah Yousuf
Status: Draft
SupercededBy: 
Version: 1.0
Area: Archive
Comments Due: 8/6/2022
Plan to implement: Yes
---

# Archive Module Version 2

This RFC proposes new features and changes for the existing Microsoft.PowerShell.Archive module.
The goal for the Archive module is to build a high-performing and maintainable module that offers high utility and works cross-platform (especially with regard to file paths).

Currently, the archive module has a number of limitations.
First, the module supports the zip32 format only.

> There is an opportunity to support additional supports by taking advanatage of new .NET APIs.

Reviewer 2: ```suggestion
There is an opportunity to support additional formats by taking advantage of new .NET APIs.

```

---
RFC:
Author: Abdullah Yousuf
Status: Draft
SupercededBy: 
Version: 1.0
Area: Archive
Comments Due: 8/6/2022
Plan to implement: Yes
---

# Archive Module Version 2

This RFC proposes new features and changes for the existing Microsoft.PowerShell.Archive module.
The goal for the Archive module is to build a high-performing and maintainable module that offers high utility and works cross-platform (especially with regard to file paths).

Currently, the archive module has a number of limitations.
First, the module supports the zip32 format only.
There is an opportunity to support additional formats by taking advantage of new .NET APIs.

Second, the module has limited performance compared to other archive software.
Writing the next version of the module in C# instead of PowerShell Script is expected to improve the overall performance of the module.

> The module has limited cross-platform support because archive entries are written in an OS-specifc way due to different characters being used as path seperators in different OSs.

Reviewer 3: I think we should pivot what is being said here. 

- There is new support in .NET for file types other than ZIP. 
- Updating the PowerShell functionality to use the new functionality in .NET there provides as chance to address some existing problems.  
- Performance is dictated by .NET - although calling from C# will reduce some of the overhead the time to do the compression or expansion is not going to change. If .NET's implementation is slower than 7Zip's , PowerShell won't fix that 
- If the new module is bundled with PowerShell **it is imperative that the module does not hinder the adoption of newer versions of PowerShell by breaking existing scripts** 
- Conversely if it fixes issues with wildcards (beyond users not knowing they need to use `-literalpath`) and improves compatibility it helps adoption of newer versions.

Author: I updated the description based on your feedback.

    As a PowerShell user,
    I can create compressed tar archives,
    so that I can reliably store and deliver content while taking up less storage space.

    As a PowerShell user,
    I can create an archive that preserves the relative path structure, so that I can keep track of which folders the contents of the archive came from and so that I can replicate the same structure on other computers.

Relative path structure refers to the hierarchial structure of a path relative to the current working directory.

    As a PowerShell user,
    I can filter what files to include in an archive,
    so that I store the necessary files only.

    As a PowerShell user,
    I can filter what files to extract from an archive,
    so that I obtain the necessary files only.

    As a PowerShell user,
    I can expand archives compressed on other OSs,
    so that I can access files in an archive without worrying about which OS it came from.

    As a PowerShell user,
    I can expand archives quickly and compress files and folders quickly,
    so that I can reduce the cost of operating my server.

> 

Reviewer 3: I would drop this last point, as performance is not under the control of the module.  Costs are only reduced by having fewer servers, or fewer people, and it's hard to see that faster compression would reduce head count or machine count.

Author: I updated this point.

Compress-Archive [-Path] <string[]> [-DestinationPath] <string> -Overwrite [-CompressionLevel {Optimal | NoCompression | Fastest | SmallestSize}] [-PassThru] [-Format {zip | tar | tgz}] [-Filter <string[]>] [-Flatten] [-WhatIf] [-Confirm] [<CommonParameters>]

Compress-Archive [-Path] <string[]> [-DestinationPath] <string> -Update [-CompressionLevel {Optimal |NoCompression | Fastest | SmallestSize}] [-PassThru] [-Format {zip | tar | tgz}] [-Filter <string[]>] [-Flatten] [-WhatIf] [-Confirm] [<CommonParameters>]

Compress-Archive [-DestinationPath] <string> -LiteralPath <string[]> [-CompressionLevel {Optimal | NoCompression | Fastest | SmallestSize}] [-PassThru] [-Format {zip | tar | tgz}] [-Filter <string[]>] [-Flatten] [-WhatIf] [-Confirm] [<CommonParameters>]

Compress-Archive [-DestinationPath] <string> -LiteralPath <string[]> -Overwrite [-CompressionLevel {Optimal | NoCompression | Fastest | SmallestSize}] [-PassThru] [-Format {zip | tar | tgz}] [-Filter <string[]>] [-Flatten] [-WhatIf] [-Confirm] [<CommonParameters>]

> Compress-Archive [-DestinationPath] <string> -LiteralPath <string[]> -Update [-CompressionLevel {Optimal | NoCompression | Fastest | SmallestSize}] [-PassThru] [-Format {zip | tar | tgz}] [-Filter <string[]>] [-Flatten] [-WhatIf] [-Confirm] [<CommonParameters>]

Author: I agree about the number of parameter sets. I think `-Collision` can be confusing to users because collision could refer to destination file collision or archive entry collision. I propose `-WriteMode { Create, Update, Overwrite}` as an alternative.

> Parameter -DestinationPath should be renamed to -Destination to match core cmdlets like Copy-Item and Move-Item. > A DestinationPath alias can be added for backwards compatibility.
How do you feel about keeping `-DestinationPath` and using `-Destination` as an alias?

> Parameter -LiteralPath should have Position = 0
> Parameter -Destination should have Position = 1
Since `-Path` has position 0, `-LiteralPath` cannot have position = 0. I agree with putting `-DestinationPath` after `-LiteralPath`.

> Compress-Archive [-DestinationPath] <string> -LiteralPath <string[]> -Update [-CompressionLevel {Optimal | NoCompression | Fastest | SmallestSize}] [-PassThru] [-Format {zip | tar | tgz}] [-Filter <string[]>] [-Flatten] [-WhatIf] [-Confirm] [<CommonParameters>]

Reviewer 3: The existing function uses `DestinationPath`. So it must be a name or an alias or existing script will break .
` -Destination` is  unambiguous as an abbreviation so there is no need to add it as an alias. 

The existing function has Path as position 0  allowing `compress-archive *.ps1`   
Swapping `path` and `literalpath`  would break existing scripts which don't formally write `-path` (and there will be many) 

Adding complexity with more parameter sets should be avoided but fixing decisions made in the original model will usually do more harm than good

> Compress-Archive [-DestinationPath] <string> -LiteralPath <string[]> -Update [-CompressionLevel {Optimal | NoCompression | Fastest | SmallestSize}] [-PassThru] [-Format {zip | tar | tgz}] [-Filter <string[]>] [-Flatten] [-WhatIf] [-Confirm] [<CommonParameters>]

Author: I updated the parameter sets. The position of `-LiteralPath` is changed so that it now comes before `-DestinationPath`. Also, the number of parameter sets is reduced.

```

```powershell
Compress-Archive -Path MyFolder -DestinationPath destination.zip -Format zip
```

A zip archive is created with the name `destination.zip`.

```powershell
Compress-Archive -Path MyFolder -DestinationPath destination.tar -Format tar
```

A tar archive is created with the name `destination.tar`.

```powershell
Compress-Archive -Path MyFolder -DestinationPath destination.tar.gz -Format tgz
```

A tar.gz archive is created with the name `destination.tar.gz`.

```powershell
Compress-Archive -Path MyFolder -DestinationPath destination.zip -Filter *.txt
```

A zip archive is created with the name `destination.zip`.

> Notice the cmdlet sees the `.zip` extension and uses the `zip` format.

Author: `-Format` is an option parameter. If `-DestinationPath` has an extension that matches a supported format (e.g., `.zip`, or `.tar`), then the cmdlet will determine the archive as the format that matches the extension.

The only files in the archive are those with the `.txt` extension.
The directory structure of `MyFolder` is maintained in the archive.

```powershell
Compress-Archive -Path MyGrandparentFolder/MyParentFolder/MyFolder -DestinationPath destination.zip 
```

A zip archive is created with the name `destination.zip`. The archive is structured as:

```
destination.zip
|---MyGrantparentFolder
    |---MyParentFolder
        |---MyFolder
            |---*
```

The archive preserves the relative structure of the input path.

> 

Reviewer 3: This will break existing scripts, which allow an absolute a relative path to be specified and will only create the `MyFolder` folder in the Zip file. 

At present `Compress-Archive -Path MyGrandparentFolder/MyParentFolder/MyFolder` treats `MyFolder` as the item to be compressed and creates a top level folder in the ZIP file,  
but `Compress-Archive -Path MyGrandparentFolder/MyParentFolder/MyFolder/*.*` puts the items in MyFolder at the top level of the zip. 

Preserving the intermediate levels is useful, but should not be the default and should be enabled by a switch.

```powershell
Compress-Archive -Path MyGrandparentFolder/MyParentFolder/MyFolder -DestinationPath destination.zip -Flatten
```

A zip archive is created with the name `destination.zip`.
The archive contains all the files in or descendents of MyFolder.
The archive **does not** retain the directory structure since `-Flatten` is specified.

> 

Reviewer 3: `-Flatten` implies all the files are in the same folder.  
Is this actually intended as a `DoNotBreakExistingScripts`  switch ?

Author: No, it is not intended as a switch that maintains backwards compatibility.

### Expand-Archive

Parameter sets:

```powershell

> Expand-Archive [-Path] <string> [[-DestinationPath] <string>] [-Overwrite] [-PassThru] [-Filter <string[]>] [-WhatIf] [-Confirm] [<CommonParameters>]
> 
> Expand-Archive [[-DestinationPath] <string>] -LiteralPath <string> [-Overwrite] [-PassThru] [-Filter <string[]>] [-WhatIf] [-Confirm] [<CommonParameters>]

Reviewer 4: ```suggestion
Expand-Archive [-Path] <string> [[-Destination] <string>] [-Overwrite] [-PassThru] [-Filter <string[]>] [-WhatIf] [-Confirm] [<CommonParameters>]

Expand-Archive [-LiteralPath] <string> [[-Destination] <string>] [-Overwrite] [-PassThru] [-Filter <string[]>] [-WhatIf] [-Confirm] [<CommonParameters>]

```

- Parameter `-DestinationPath` should be renamed to `-Destination` to match core cmdlets like `Copy-Item` and `Move-Item`. A `DestinationPath` alias can be added for backwards compatibility.

- Parameter `-LiteralPath` should have `Position = 0`

- Parameter `-Destination` should have `Position = 1`

- Parameter `-Overwrite` should be named `-Force` to match core cmdlets or if that is unacceptable a `Force` parameter alias added.

Author: > Parameter -DestinationPath should be renamed to -Destination to match core cmdlets like Copy-Item and Move-Item. A DestinationPath alias can be added for backwards compatibility.
> Parameter -LiteralPath should have Position = 0
> Parameter -Destination should have Position = 1

See my comment above. 

> Parameter -Overwrite should be named -Force to match core cmdlets or if that is unacceptable a Force parameter alias added.

PowerShell is moving away `-Force` because it is vague. `-Overwrite` is more specific and preferred.

Reviewer 4: @ayousuf23 the only time `Force` is an issue is when it's not instantly clear what it will do and if it does more than one thing. In this case that is neither the concern or issue. `Install-Module` is the perfect example of bad `Force` design where it does too many things, reinstalls the module, disables certain checks, and suppresses prompts. PowerShell design works best when the commands are consistent. `Force` has been used since the beginning and does have some issues like I described above but if the user has been using PowerShell for any length of time, they will first try `Force`. As such if the PowerShell team is unwilling to change the name then be nice to your long time users by adding a `Force` alias.

P.S. None of the current built-in commands have `Overwrite` as a parameter or alias. There is however, a `NoOverwrite` alias for a few `NoClobber` parameters. I'd recommend the team to add `Overwrite` aliases to existing `Force` parameters where it makes sense to be more consistent if `Overwrite` will be the new standard parameter for this behavior.

Reviewer 3: @ayousuf23   Although there may be a desire to reduce the use of the `-Force` switch, it is well understood (don't confirm, do overwrite) and is embedded in many existing scripts. 

The way to resolve the conflict between compatibility and fixing things which the original design didn't get quite right  would be to use **new command names.**  
 We don't take an archive and compress it.  Really the command should be `New-Archive`, and have `Add-ArchiveFile`, `Remove-ArchiveFile` and `Update-ArchiveFile`  commands for changing an existing file and `Show-ArchiveFile`  for listing its contents.  
`Expand` doesn't allow a subset of files to be extracted, `copy-archiveFile` would be ideal for that task but I'm not sure if that is as good a name for the common use of expand everything, but (again) changing the name sidesteps the problem of breaking existing scripts.

Author: > The way to resolve the conflict between compatibility and fixing things which the original design didn't get quite right would be to use new command names.

Good idea!

> We don't take an archive and compress it. Really the command should be New-Archive, and have Add-ArchiveFile, Remove-ArchiveFile and Update-ArchiveFile commands for changing an existing file and Show-ArchiveFile for listing its contents.

This is an interesting idea. I need to think about it some more before I get back to you.

> Expand doesn't allow a subset of files to be extracted, copy-archiveFile would be ideal for that task but I'm not sure if that is as good a name for the common use of expand everything, but (again) changing the name sidesteps the problem of breaking existing scripts.

We can add a parameter to `Expand-Archive` that allows a subset of files to be extracted. This way it won't break existing scripts and it can still be used.

Reviewer 5: I'd rather this module keep using `-Force` for compatibility's sake.

Example: `Compress-Archive -Path MyFolder -DestinationPath destination.tar -Format zip`

The zip format is chosen for the archive since the `-Format` parameter takes precedence over the extension of destination.tar.
A warning is reported to the user about the mismatch between the archive extension and chosen format value.

Example: `Compress-Archive -Path MyFolder -DestinationPath destination`

The zip format is chosen for the archive by default.
A warning, notifying the user that value of `-DestinationPath` has a missing extension and that the zip format is chosen as the archive format by default, is reported to the user.
Note that `.zip` is not appended to the archive name.

#### **Relative Path Structure Preservation**

When valid path(s) is supplied to the `-Path` or `-LiteralPath` parameter, the relative structure of the path is preserved as long as the path is not absolute, and does not contain ~ or .. (the path can contain other wildcard characters such as *, [, ], ?, and `).

> 

Reviewer 3: See above that preserving the path will be a breaking change for existing scripts 
In a PATH parameter  `File[12]`  means `File1` or `File2` in a literal path [12] is allowed to mean the twelfth file of a a set. It is important that users can still use (e.g)  "Log[0-9]*.txt"  to include files with "log" a digit and anything else. but exclude "Logarithms.txt"  - I think this probably understood, but it does not come through reading the text.

Author: The `-LiteralPath` parameter behaves the same as before in regard to interpreting wildcard characters literally (i.e., not expanding wildcard characters).

    As a PowerShell user,
    I can create compressed tar archives,
    so that I can reliably store and deliver content while taking up less storage space.

    As a PowerShell user,
    I can create an archive that preserves the relative path structure, so that I can keep track of which folders the contents of the archive came from and so that I can replicate the same structure on other computers.

Relative path structure refers to the hierarchial structure of a path relative to the current working directory.

    As a PowerShell user,
    I can filter what files to include in an archive,
    so that I store the necessary files only.

    As a PowerShell user,
    I can filter what files to extract from an archive,
    so that I obtain the necessary files only.

    As a PowerShell user,
    I can expand archives compressed on other OSs,
    so that I can access files in an archive without worrying about which OS it came from.

    As a PowerShell user,
    I can expand archives quickly and compress files and folders quickly,
    so that I can reduce the cost of operating my server.

> 

Reviewer 3: There is a design flaw with Zip which goes back to its origins on MS DOS. It stores the date/time as displayed - i.e. as local time, not UTC time. And when the file is extracted the time is set as local time. This means if compress and expand happen on machines with different time zone settings the time stamp on the files changes. Changing the file format to fix this is absolutely out of scope, however a -SourceTimeZone parameter could allow the unzip operation to create correctly dated files.

Author: The problem with `-SourceTimeZone` is that it seems like it is only needed for Zip and not other archive formats. How can we make this work for other formats as well?

Reviewer 3: If other formats store the as "wall time time with the offset from utc" - equivalent to `Get-Date -Format "o"`  or always use the same timezone (UTC , Pacific, whatever),  then the best course is either to show a warning and ignore any time adjustment parameter, or possibly to apply it only if a `-Force` or a similar switch is used.   Zip has never become time zone aware, because either it would mean new files couldn't be unzipped by old programs, or files would get different dates depending on which version unzipped them. 

Applying a fixed offset to all files doesn't work, because daylight savings time moves the clock in opposite directions in the Northern and Southern Hemisphere and is applied at different dates.  .NET timezones are smart enough to know that what offset applies in particular timezone at a given time.  (I wrote about it :-) https://jhoneill.github.io/2022/03/07/Timezones.html )

Author: This is a good idea. I think we can keep this on the backlog and add this feature as time permits.

Example: `Compress-Archive -Path Documents -DestinationPath destination.zip -Flatten` creates an archive which only contains the files in or descended from `Documents` and these files are the top-level items.

> The `-Flatten` parameter can be used with `-WriteMode Update`.

Reviewer 3: `-Update`  is used in existing scripts, so is needed to avoid breaking them.

`-Writemode` is better , so the code needs to have an `Update` switch with the DontShow flag set, and if specified this will set the value  Writemode.

Reviewer 4: The same goes with `-Force` to avoid a breaking change a hidden force parameter that sets `WriteMode` to `Overwrite`.
